DOCS: Read The Docs and doxygen integration

shamisp opened the pull request:
## What
Read The Docs and doxygen integration

## Why ?
Generates very nice website for our API automatically: https://ucx-shamisp.readthedocs.io/en/latest/api.html

## How ?
Using sphinx + breath + Doxygen XML

swx-jenkins3 commented:
Can one of the admins verify this patch?

shamisp commented:
* Depends on #6841. Once merged I will rebase.
* Readthedocs default built will fail in testing since I have to update configuration to pick up requirements file. It will be done once the PR is merged. For now use the above URL for preview.

shamisp commented:
ok to test

yosefe commented:
Few general comments:
- How can we have API doc per released version (so that unreleased API changes in master branch will not be in the main doc page)?
- Is there an index of all functions like in doxygen html?

shamisp commented:
> Few general comments:
> 
> * How can we have API doc per released version (so that unreleased API changes in master branch will not be in the main doc page)?

* The proper way of doing this is to redirect readthedocs from master to latest release. 
* Anything else will not look pretty - explicit checkout with git into subdirectory and running doxygen over the external tree. 

> * Is there an index of all functions like in doxygen html?

Unfortunately no. We can create some like this by listing functions manually but it is hard to maintain for a large API.

shamisp commented:
@yosefe seems like sputnik failures are irrelevant and the rtd is expected

yosefe commented:
> @yosefe seems like sputnik failures are irrelevant and the rtd is expected

why is it expected? does it mean RTD will not build properly now?

shamisp commented:
@yosefe I have to update RTD configuration to point to the new requirements file. Essentially once it is merged, I have to update RTD config. On my personal RTD it is already updated.

yosefe commented on docs/source/api.rst:
```diff
@@ -0,0 +1,56 @@
+
+Worker routines
+*******************
+.. doxygengroup:: UCP_WORKER
```
can we build this token list automatically somehow?

shamisp replied:
They do have an external tool to do it, but so far I didn't find it very reliable (causes random segfaults). I would say for now hardcoded is better but in a future we can potentially improve on this.

yosefe replied:
ok

yosefe commented on docs/source/conf.py:
```diff
@@ -62,5 +63,56 @@ def substituteVersion(app, docname, source):
+    import os
+
+    read_the_docs_build = os.environ.get('READTHEDOCS', None) == 'True'
+
```
minor: remove space line

yosefe commented on docs/source/index.rst:
```diff
@@ -69,13 +71,6 @@ and run a simple client/server example:
     ------------------------------
 
 
-Documentation
```
why removed? let's keep the link to PDF and HTML, sometimes it's more convenient

shamisp replied:
I moved it to the head of API section.

yosefe replied:
i wonder if we can put it to a more obvious place in the main page?

shamisp replied:
IMHO - the original location was bad. I think placing this at the top of API section is a good place. People that care about API will click on the menu item ( on the left side) and the first thing they will see is the reference to full spec.

yosefe replied:
okay

yosefe commented on docs/source/api.rst:
```diff
@@ -0,0 +1,56 @@
+.. See file LICENSE for terms.
+..
+
+API
```
maybe "API specification"?

shamisp replied:
Since this is not a full document, I avoided to call it specification. In the top of the page we do provide references to the full spec.

yosefe replied:
maybe "API reference" then?

shamisp replied:
I was thinking UCX API or just API. I'm not sure what "reference" adds to the context. Maybe "API Documentation" but IMHO documentation part is kind of obvious...

yosefe replied:
okay, since there is no better suggestion for now let's keep it as is. can always improve later.

yosefe commented:
> @yosefe I have to update RTD configuration to point to the new requirements file. Essentially once it is merged, I have to update RTD config. On my personal RTD it is already updated.

ok

shamisp commented:
@yosefe I made few additional changes:
* Index generation for files and structs (doxygen does not support functions) - https://ucx-shamisp.readthedocs.io/en/latest/api.html
* Reduced number of request to GitHub (it does not like it...)
* Moved configuration to yaml
* Removed cross reference from UCP.H to UCT, otherwise it breaks if UCT API is not generated

shamisp commented on docs/source/conf.py:
```diff
@@ -44,10 +53,19 @@
 # UCX custom configuration
 #
 
+# For local testing use return 'v1.10.1', otherwise github may block you for
```
Yes, I will clarify. Github essentially has DDOS protection, so when you build locally a lot , you want just overwrite the value.

shamisp replied:
done

shamisp commented on docs/source/conf.py:
```diff
@@ -44,10 +53,19 @@
+
+    global cached_version
+    if cached_version == '':
+        request = requests.get('https://api.github.com/repos/openucx/ucx/releases/latest')
```
I didn't find one ...

shamisp replied:
https://docs.readthedocs.io/en/stable/builds.html

shamisp commented:
> Also, for some reason cannot open link https://ucx-shamisp.readthedocs.io/en/latest/api.html (SSL/HTTPS error)

Works for me (phone, laptop, etc.). You can check this out locally and run `make html` to build local revision.

shamisp commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,19 @@
+# .readthedocs.yaml
```
done

shamisp commented:
@yosefe squash or leave it as it is ?

yosefe reviewed (approved):
pls squash
